common/src/main/java/io/netty/util/internal/HashCodeGenerator.java

```diff
@@ -0,0 +1,53 @@
+/*
+ * Copyright 2015 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.util.internal;
+
+/**
+ * Provides the ability to generate a hash code for array based objects.
+ */
+public interface HashCodeGenerator {
+    /**
+     * Same as calling {@link #hashCode(byte[], int, int)} as {@code hashCode(bytes, 0, bytes.length)}.
+     */
+    int hashCode(byte[] bytes);
+
+    /**
+     * Generate a hash code from the [{@code startPos}, {@code endPos}) subsection of {@code bytes}.
+     */
+    int hashCode(byte[] bytes, int startPos, int endPos);
+
+    /**
+     * Same as calling {@link #hashCodeAsBytes(char[], int, int)} as {@code hashCode(bytes, 0, bytes.length)}.
+     */
+    int hashCodeAsBytes(char[] bytes);
```

Comment:
I will add additional comments.  The reason for the name difference is the same rational as https://github.com/netty/netty/pull/3694#discussion_r29120238.  The MSB of each `char` is ignored.